### PR TITLE
Adding IPv6 rule support in Nephe

### PR DESCRIPTION
**Pull request by reachjainrahul**
## Description
Handle IPv6 addresses configured on Antrea NetworkPolicy

## Changes 
With this checkin, Nephe controller will accept IPv6 IP blocks in Antrea NetworkPolicy and will apply them in cloud. Also, VirtualMachine Object will now include both IPv4 and IPv6 addresses in the network interface field.

### CODE REVIEW

**Comment by reachjainrahul** [comment]
/nephe-test-e2e-azure, /nephe-test-e2e-aks /nephe-test-e2e-eks

File: pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go [hunk 1]
```diff
@@ -46,30 +46,41 @@ func ec2InstanceToInternalVirtualMachineObject(instance *ec2.Instance, vpcs map[
...
 		}
+		// Parsing Amazon allocated IPv6 addresses for now. They are treated as Public IPs.
+		// TODO: Re-check the parsing for private IPv6 IP block.
+		if len(nwInf.Ipv6Addresses) > 0 {
+			for _, ipv6Address := range nwInf.Ipv6Addresses {
+				ipAddress := runtimev1alpha1.IPAddress{
+					AddressType: runtimev1alpha1.AddressTypeExternalIP,
+					Address:     *ipv6Address.Ipv6Address,
+				}
+				ipAddressObjs = append(ipAddressObjs, ipAddress)
```
Comment (Anandkumar26) [inline pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go:77]: For the agented, are the Externalnodes getting created correctly?

File: pkg/cloudprovider/cloudapi/aws/aws_converters.go [hunk 2]
```diff
@@ -45,28 +45,39 @@ func convertToIPPermissionPort(port *int, protocol *int) (*int64, *int64) {
...
+	var ipv6Ranges []*ec2.Ipv6Range
 	if len(ips) == 0 && !ruleHasGroups {
 		ipRange := &ec2.IpRange{
 			CidrIp:      aws.String("0.0.0.0/0"),
 			Description: description,
 		}
-		ipRanges = append(ipRanges, ipRange)
-		return ipRanges
+		ipv4Ranges = append(ipv4Ranges, ipRange)
+		return ipv4Ranges, ipv6Ranges
```
Comment (shenmo3) [inline pkg/cloudprovider/cloudapi/aws/aws_converters.go:58]: This means allow all right? So shouldn't an IPv6 `::` range also be added?

File: pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go [hunk 3]
```diff
@@ -46,30 +46,41 @@ func ec2InstanceToInternalVirtualMachineObject(instance *ec2.Instance, vpcs map[
...
 					}
-					ipAddressCRDs = append(ipAddressCRDs, ipAddressCRD)
+					ipAddressObjs = append(ipAddressObjs, ipAddressCRD)
 				}
 			}
 		}
+		// Parsing Amazon allocated IPv6 addresses for now. They are treated as Public IPs.
+		// TODO: Re-check the parsing for private IPv6 IP block.
+		if len(nwInf.Ipv6Addresses) > 0 {
+			for _, ipv6Address := range nwInf.Ipv6Addresses {
```
Comment (shenmo3) [inline pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go:72]: is the if block needed? a for block is sufficient isn't it?

File: pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go (hunk 1 above)
Comment (reachjainrahul) [reply pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go:77]: Yes..

File: pkg/inventory/inventory.go [hunk 4]
```diff
@@ -248,7 +250,35 @@ func (inventory *Inventory) GetVmByKey(key string) (*runtimev1alpha1.VirtualMach
...
+
+// compareVirtualMachineObjects compare if two virtual machine objects are the same. Return true if same.
+func (i Inventory) compareVirtualMachineObjects(cached, discovered runtimev1alpha1.VirtualMachineStatus) bool {
+	// 1. Check if objects are same.
+	if reflect.DeepEqual(cached, discovered) {
+		return true
+	}
+	// 2. Check if NetworkInterface field differ.
+	if reflect.DeepEqual(cached.NetworkInterfaces, discovered.NetworkInterfaces) {
+		return false
```
Comment (shenmo3) [inline pkg/inventory/inventory.go:266]: shouldn't this return true if nics are same?

Comment (reachjainrahul) [reply pkg/inventory/inventory.go:266]: Azure returns nic address not in particular order. So we sort them and then compare

File: pkg/cloudprovider/cloudapi/aws/aws_converters.go (hunk 2 above)
Comment (reachjainrahul) [reply pkg/cloudprovider/cloudapi/aws/aws_converters.go:58]: you are right.. let me add that

File: pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go (hunk 3 above)
Comment (reachjainrahul) [reply pkg/cloudprovider/cloudapi/aws/aws_crd_helpers.go:72]: its generally a good practice, but not must..

**Review by shenmo3 — APPROVED** [review]

**Pull request merged**